### PR TITLE
IPStrategy for selecting IP in whitelist

### PR DESCRIPTION
**Pull request by Juliens**
### What does this PR do?

- It handles ForwardedHeaders in a dedicated middleware.
- ForwardedHeaders are now not trusted by default (delete each forwarded headers)
- Create an new IPStrategy to select the clientIP
- Whitelist use this new IPStrategy

### Motivation
More secure way to handle forwarded headers and to handle the whitelist.

### More

- [X] Added/updated tests
- [X] Added/updated documentation

### CODE REVIEW

File: configuration/router/internal_router.go [hunk 1]
```diff
@@ -16,9 +16,15 @@ func NewInternalRouterAggregator(globalConfiguration configuration.GlobalConfigu
...
-			globalConfiguration.EntryPoints[entryPointName].WhiteList.UseXForwardedFor)
+		ipStrategy := globalConfiguration.EntryPoints[entryPointName].ClientIPStrategy
+		if globalConfiguration.EntryPoints[entryPointName].WhiteList.IPStrategy != nil {
+			ipStrategy = globalConfiguration.EntryPoints[entryPointName].WhiteList.IPStrategy
+		}
+		strategy, err := ipStrategy.Get()
+		if err != nil {
+			log.Fatalf("Error creating whitelist middleware: %s", err)
+		}
+		ipWhitelistMiddleware, err := middlewares.NewIPWhiteLister(globalConfiguration.EntryPoints[entryPointName].WhiteList.SourceRange, strategy)
```
Comment (mmatur) [inline configuration/router/internal_router.go]: Could you please add a new line before

File: configuration/router/internal_router.go [hunk 2]
```diff
@@ -16,9 +16,15 @@ func NewInternalRouterAggregator(globalConfiguration configuration.GlobalConfigu
...
 
 	if globalConfiguration.EntryPoints[entryPointName].WhiteList != nil {
-		ipWhitelistMiddleware, err := middlewares.NewIPWhiteLister(
-			globalConfiguration.EntryPoints[entryPointName].WhiteList.SourceRange,
-			globalConfiguration.EntryPoints[entryPointName].WhiteList.UseXForwardedFor)
+		ipStrategy := globalConfiguration.EntryPoints[entryPointName].ClientIPStrategy
+		if globalConfiguration.EntryPoints[entryPointName].WhiteList.IPStrategy != nil {
+			ipStrategy = globalConfiguration.EntryPoints[entryPointName].WhiteList.IPStrategy
+		}
+		strategy, err := ipStrategy.Get()
```
Comment (mmatur) [inline configuration/router/internal_router.go]: Could you please add a new line before

File: integration/basic_test.go [hunk 3]
```diff
@@ -339,3 +340,80 @@ func (s *SimpleSuite) TestMultipleProviderSameBackendName(c *check.C) {
...
+			expectedStatusCode: 200,
+		},
+	}
+
+	for _, test := range testCases {
+		req := httptest.NewRequest(http.MethodGet, "http://127.0.0.1:8000", nil)
+		req.Header.Set("X-Forwarded-For", test.xForwardedFor)
+		req.Host = test.host
+		req.RequestURI = ""
+		err = try.Request(req, 1*time.Second, try.StatusCodeIs(test.expectedStatusCode))
```
Comment (mmatur) [inline integration/basic_test.go]: Could you please add a new line before

File: integration/fixtures/access_log_config.toml [hunk 4]
```diff
@@ -17,7 +17,8 @@ checkNewVersion = false
       entryPoint = "http"
   [entryPoints.httpWhitelistReject]
     address = ":8002"
-    whiteListSourceRange = ["8.8.8.8/32"]
+  [entryPoints.httpWhitelistReject.whiteList]
+    SourceRange = ["8.8.8.8/32"]
```
Comment (mmatur) [inline integration/fixtures/access_log_config.toml]: Could you rename to `sourceRange`

File: integration/fixtures/simple_whitelist.toml [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+[entryPoints]
+  [entryPoints.http]
+  address = ":8000"
+  [entryPoints.http.ForwardedHeaders]
+    insecure=true
+  [entryPoints.http.ClientIPStrategy]
+    depth=2
+
+
```
Comment (mmatur) [inline integration/fixtures/simple_whitelist.toml]: Could you remove this line please

File: integration/resources/compose/whitelist.yml [hunk 6]
```diff
@@ -0,0 +1,34 @@
+noOverrideWhitelist:
+  image: emilevauge/whoami
```
Comment (mmatur) [inline integration/resources/compose/whitelist.yml]: Could you change image to `containous/whoami`

File: integration/resources/compose/whitelist.yml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  labels:
+    - traefik.enable=true
+    - traefik.port=80
+    - traefik.frontend.rule=Host:no.override.whitelist.docker.local
+    - traefik.frontend.whiteList.sourceRange=8.8.8.8
+
+overrideIPStrategyRemoteAddrWhitelist:
+  image: emilevauge/whoami
```
Comment (mmatur) [inline integration/resources/compose/whitelist.yml]: Could you change image to `containous/whoami`

File: integration/resources/compose/whitelist.yml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  labels:
+    - traefik.enable=true
+    - traefik.port=80
+    - traefik.frontend.rule=Host:override.remoteaddr.whitelist.docker.local
+    - traefik.frontend.whiteList.sourceRange=8.8.8.8
+    - traefik.frontend.whiteList.ipStrategy=true
+
+overrideIPStrategyDepthWhitelist:
+  image: emilevauge/whoami
```
Comment (mmatur) [inline integration/resources/compose/whitelist.yml]: Could you change image to `containous/whoami`

File: integration/resources/compose/whitelist.yml [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  labels:
+  - traefik.enable=true
+  - traefik.port=80
+  - traefik.frontend.rule=Host:override.depth.whitelist.docker.local
+  - traefik.frontend.whiteList.sourceRange=8.8.8.8
+  - traefik.frontend.whiteList.ipStrategy.depth=3
+
+overrideIPStrategyExcludedIPsWhitelist:
+  image: emilevauge/whoami
```
Comment (mmatur) [inline integration/resources/compose/whitelist.yml]: Could you change image to `containous/whoami`

File: ip/checker.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+			checker.authorizedIPs = append(checker.authorizedIPs, &ipAddr)
+		} else {
+			_, ipAddr, err := net.ParseCIDR(ipMask)
+			if err != nil {
+				return nil, fmt.Errorf("parsing CIDR trusted IPs %s: %v", ipAddr, err)
+			}
+			checker.authorizedIPsNet = append(checker.authorizedIPsNet, ipAddr)
+		}
+	}
+	return &checker, nil
```
Comment (mmatur) [inline ip/checker.go]: Could you please add a new line

File: ip/checker.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+	return nil
+}
+
+// Contains checks if provided address is in the trusted IPs
+func (ip *Checker) Contains(addr string) (bool, error) {
+	if len(addr) <= 0 {
+		return false, errors.New("empty IP address")
+	}
+	ipAddr, err := parseIP(addr)
+	if err != nil {
```
Comment (mmatur) [inline ip/checker.go]: Could you please add a new line before?

File: ip/checker_test.go [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+				require.NoError(t, err)
+			} else {
+				require.Error(t, err)
+			}
+		})
+	}
+}
+
+func TestNew(t *testing.T) {
+	cases := []struct {
```
Comment (mmatur) [inline ip/checker_test.go]: Could you please rename `cases` by `testCases`

File: ip/checker_test.go [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+			},
+			errMessage: "",
+		},
+	}
+
+	for _, test := range cases {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			t.Parallel()
+			ipChecker, err := NewChecker(test.trustedIPs)
```
Comment (mmatur) [inline ip/checker_test.go]: Could you please add a new line before

File: ip/checker_test.go [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+					assert.Equal(t, expected.IP, actual.IP)
+					assert.Equal(t, expected.Mask.String(), actual.Mask.String())
+				}
+			}
+		})
+	}
+}
+
+func TestContainsIsAllowed(t *testing.T) {
+	cases := []struct {
```
Comment (mmatur) [inline ip/checker_test.go]: Could you please rename `cases` by `testCases`

File: ip/strategy.go [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+		return ""
+	}
+	xff := req.Header.Get(xForwardedFor)
+	xffs := strings.Split(xff, ",")
+	for i := len(xffs) - 1; i >= 0; i-- {
+		if contain, _ := s.Checker.Contains(xffs[i]); !contain {
+			return xffs[i]
+		}
+	}
+	return ""
```
Comment (mmatur) [inline ip/strategy.go]: Could you please add a new line before

File: server/server_middlewares.go [hunk 16]
```diff
@@ -277,14 +278,20 @@ func (s *Server) buildRedirectHandler(srcEntryPointName string, opt *types.Redir
...
+
+	if whiteList.IPStrategy != nil {
+		ipStrategy = whiteList.IPStrategy
+	}
+
+	strategy, err := ipStrategy.Get()
+	if err != nil {
+		return nil, err
+	}
+	return middlewares.NewIPWhiteLister(whiteList.SourceRange, strategy)
```
Comment (mmatur) [inline server/server_middlewares.go]: Could you please add a new line before

**Comment by jbdoumenjou** [comment]
LGTM

**Pull request merged**